### PR TITLE
add sl-extended

### PR DESCRIPTION
**Pull request by tempfactorio**
I tested the current redmew map with these added files and it worked.

Feel free to remove features.

source: http://bit.ly/sl-extended3

https://forums.factorio.com/viewtopic.php?f=94&t=39562

features: 

**Custom command:**

- /slap <player> - Slap a player.
- /slay <player> - Slay a player. (admin only)
- /logistics <endgame,> - Sets logistics requests slots.
- /inv <-100,100> - Decrease,increase character inventory bonus.
- /req <-18,18> - Decrease,increase character logistic slots bonus.
- /hp <-1000,1000> - Decrease,increase character health bonus.
- /toolbar <-3,3> - Remove,add a toolbar bonus.
- /run <-3,3> - Decrease,increase character running speed bonus.
- /zoom <1,100> - Sets the player's zoom-level.
- /acquire - Get all your requests in your inventory. (admin only)
- /robotspeedresearch - Increases the robot speed research level. (admin only) 

**Start items:** 

- 25 iron plates
- 25 stone

**Spawn items:**

- construction robots (and supporting armor)
- steel axe
- deconstruction planner

**Character optimizations:**

- character long reach
- character mining speed: 50%
- character crafting speed: 150%

**Scheduler:** 

**_Automatically sets train schedule:_**

- Open/click a train (locomotive, cargo-wagon, fluid-wagon)
- Press "Scheduler"-button (top left)
- Train Stop:
- Select station name
- Select type <"loading","unloading"> (see note)
- Press "add"-button 
- (repeat 3)
- Press "finish"-button to save
- note: Autodetected type by station name
- load
- iron*, copper*, coal*, stone*, uranium*, oil* (eg. 'Iron 05', 'Copper 02', 'Uranium 05', 'Oil 01', 'Oil 02', ..)
- rocket*science* (eg. 'Rocket 01 Space Science','Rocket 02 Space Science', ..)
- base*acid* (eg. 'Base Acid')
- depot* (eg. 'Depot Iron', 'Depot Copper', 'Depot Coal', 'Depot Uranium', 'Depot Stone', 'Depot Space Science', 'Depot Oil', 'Depot Acid', ..)
- unload
- *unload* (eg. 'Depot Unload', 'Depot Oil Unload', 'Depot Acid Unload', ..)
- acid* (eg. 'Acid 01', 'Acid 09', ..)
- base* (eg. 'Base Iron', 'Base Copper', 'Base Space Science', 'Base Coal', 'Base Uranium', 'Base Stone', 'Base Oil', ..)
- science* (eg. 'Science 01 Iron', 'Science 01 Copper', 'Science 01 Space Science', 'Science 01 Oil', ..)
- rocket* (eg. 'Rocket 01 Iron', 'Rocket 01 Copper', 'Rocket 01 Oil', ..)

**Chest**: Clicking it while holding a blueprint(book) will place all the required items in the logistic requester chest. (note: removes old requests, Not more than 12 slots)
**Character**: Clicking it while holding a blueprint(book) will place all the required items in the logistics requester slots. (note: removes old requests, Adds more slots if required)
**Upgrade Planner ++**: instant, without robots

- While holding a deconstruction planner:

1. Tree removal
2. Rock removal
3. Pickup dropped items
4. Upgrade planner
5. Turret removal: remove turrets and there ammo instantly.

- While holding a tile of concrete or stone-brick (size 1x1):

1. Dig: 1x1 tile > 7x7 water

**Auto Deconstruct**: mark idle miners for deconstruction
**Itemcount**: show the total amount of items you are holding in your cursor and not just the amount in that single stack.
**Remove bloated map decorations** to shrink the save file for faster multiplayer (eg. grass, fish, ..)
**Autofill**: automatically adds fuel and ammo from inventory to the entities you build
**Keeping blueprint knowledge** after manually placing an item (eg. blueprint of a assembling machine building Gears. When a player places the assembler it automatically sets the recipe.)
**Show a message when a rocket is launched**, seconds since last launch and launch counter: eg. "[568 S] A rocket silo launched satellite 19"
**Auto capitalize** renamed entities. (the first letter of each word)

### CODE REVIEW

**Comment by TWLTriston** [comment]
First off, this is great, thanks for contributing to the repo! Thoughts:

Generally, I like QoL improvements, but not to the edge of cheating -- specifically have concerns about:
1) Long Reach --- Can this be a regular only feature, or must it apply to all users? Having this would make griefing easier to do, and harder to find.
2) Completely overhauling train schedule, while nifty, would add a layer of learning that many users would be unwanting / unable to fully comprehend
3) Instant upgrading -- does that take into account what's in your inventory, or free belts?
4) /hp -- Feels cheaty

Code style comments:
A) File structure -- would want to keep the SL features into their own area, and reduce the on/off to a single command in the map_layout.lua file such that we can include / exclude easily. These become a core feature, however I don't want to have massive changes in order to enable/disable.
B) Auto deconstruct was recently added by Val -- would want to ensure we don't have a conflict there.
C) Need to verify if upgradeplanner_on_marked_for_deconstruction and the non-regular check for deconstruction will still stop malicious users.
D) Suggestion on dig_positions, use two loops to go -3,+3 x/y to reduce code length and improve readability
E) sl_utils -- is that an empty file?

**Comment by tempfactorio** [comment]
1. **Longreach:** 
It could be a regular and up feature. Currently it's called at 2 events (join & respawn)

2. **Scheduler:** 
The vanilla way is still available. It's only faster to use this scheduler. You can preset your favorite conditions.

3. **Upgradeplanner:**
It only upgrades if you have the items in your inventory, otherwise you get a floating message. (something ...  'not enough blue belts')

4. **/hp:**
you're right

D. **dig_positions** could be improved. But keep in mind that the order is important.

E. **sl_utils.lua** should not be empty. I'll try to add the content.

File: sl_utils.lua [hunk 1]
```diff
@@ -1 +1,957 @@
...
+end
+
+function formattime(ticks)
+  local seconds = ticks / 60
+  local minutes = math.floor((seconds)/60)
+  local seconds = math.floor(seconds - 60*minutes)
+  return string.format("%dm:%02ds", minutes, seconds)
+end
+
+-- Simple function to get total number of items in table
```
Comment (TWLTriston) [inline sl_utils.lua]: How is this different from: return #T

File: sl_utils.lua [hunk 2]
```diff
@@ -1 +1,957 @@
...
+-- Slay a player.
+function slay(targetPlayer, fromPlayer)
+    -- check if fromPlayer has the rights to slap playername..
+    if targetPlayer and targetPlayer.valid and targetPlayer.connected then
+        slSaysAll(fromPlayer.name .. " slayed " .. targetPlayer.name)
+        targetPlayer.character.die()
+    end
+end
+
+-- Give player these default items.
```
Comment (TWLTriston) [inline sl_utils.lua]: Will want to discuss balance on this item list -- leaning more towards power armor, 1 robo port and 10 robots, just to ensure there's something to expand towards

**Comment by Valansch** [comment]
Hi SI!
Thanks for contributing. Didn't know you where planning on adding it.

Can you please make this multiple pullrequests? Makes discussing this easier and this is alot of code.

While i have some comments on the code and some more specific questions about some of the issues, i'll reserve them for later / the seperate PR. Since i think we have to talk about what we actually need and/or want.

Some of the things can be integrated into our existing code, some of it we can use if adapted, and some of it we don't need (and want).

So here are my personal thoughts on the features:

- Features that need to be adapted
  - /slap - Slap a player.
We can integrate this into the pokes directly.

  - /slay - Slay a player. (admin only)
We can intergrate this into /kill (/kill would kill the admin /kill player would kill the player)

  -  Show a message when a rocket is launched, seconds since last launch and launch counter: eg. "[568 S] A rocket silo launched satellite 19"
 This is a good idea, but the text should only say something like "A rocket silo launched satellite 19" without any tags


- Features that are too cheaty:
  - /logistics <endgame,> - Sets logistics requests slots.
  - /inv <-100,100> - Decrease,increase character inventory bonus.
  - /req <-18,18> - Decrease,increase character logistic slots bonus.
  - /hp <-1000,1000> - Decrease,increase character health bonus.
  - /toolbar <-3,3> - Remove,add a toolbar bonus.
  - /run <-3,3> - Decrease,increase character running speed bonus.
  - /acquire - Get all your requests in your inventory. (admin only)
  - /robotspeedresearch - Increases the robot speed research level. (admin only)
  - Spawn items
    - construction robots (and supporting armor)
    - steel axe
    - deconstruction planner
   - Upgrade Planner ++
   - (Upgrade Planner)
     We can talk about this one but i personally still think its too cheaty. Maybe we ask other admins for their opinion
   - character long reach
   - While holding a deconstruction planner:
     - Tree removal
     - Rock removal
     - Turret removal: remove turrets and there ammo instantly.
   This would also interfere with our anti griefer script

- Features we don't need / interfere with other features we have
  - Remove bloated map decorations
  Any map related things are handles by the map code itself and should only be added as additional modules (as "combined module" or "misc module")
  - Auto capitalize renamed entities. (the first letter of each word)
    So instead of *heXXor* it would be *Hexxor*? Why spell heXXors username differently then he intended to? :(

- Features we can implement right away
  - /zoom <1,100> - Sets the player's zoom-level.
This seems cool.
  - Pickup dropped items
  - Itemcount
  - Autofill


- Things we already have:
  - Start items
  - Auto Deconstruct
 
- Features i don't know why we need
  - Character optimizations
    - character mining speed: 50%
    - character crafting speed: 150%
   - Scheduler
      I don't think having the scheduler is a good idea, it will just confused people and only half the people will use it and the other one will rage. And since it takes up additional gui space in the top left i want to be carefull about not overbloating the gui with things that arent actually used. Maybe we can implement the scheduler as a command instead?


- Things i don't understand what they do:
    - Chest: Clicking it while holding a blueprint(book) will place all the required items in the logistic requester chest. (note: removes old requests, Not more than 12 slots)
    - Character: Clicking it while holding a blueprint(book) will place all the required items in the logistics requester slots. (note: removes old requests, Adds more slots if required)

   - While holding a tile of concrete or stone-brick (size 1x1):
    Dig: 1x1 tile > 7x7 water
   So i cant draw anymore? 

    - Keeping blueprint knowledge after manually placing an item (eg. blueprint of a assembling machine building Gears. When a player places the assembler it automatically sets the recipe.)

**Comment by tempfactorio** [comment]
I could show all features in game.

eg.
/**logistics** <endgame,> - Sets logistics requests slots.
This is not a cheat because it only sets the slots to my predefined end game. Could be a feature you don't need.
everything in the **upgrade planner ++** uses the **deconstruction planner** (except dig). pickup dropped items, upgrade, tree, rock, 
**rocket launch message**: the 'tag' is actually the seconds since your last launch.

**Hidden**: in sl-extended most gui is hidden unless all conditions are met (visible/invisible/hidden)

**Pull request closed without merging**